must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java

```diff
@@ -71,6 +75,21 @@ protected boolean skipReceiverSubtypeCheck(
     return true;
   }
 
+  /**
+   * Skip assignment checks for try-with-resources variables, because they're always @MustCall({}).
+   */
+  @Override
+  protected void commonAssignmentCheck(
+      Tree varTree,
+      ExpressionTree valueExp,
+      @CompilerMessageKey String errorKey,
+      Object... extraArgs) {
+    if (TreeUtils.elementFromTree(varTree).getKind() == ElementKind.RESOURCE_VARIABLE) {
```

Comment:
@kelloggm do we still need this change?  If we do, we should update the Javadoc as it's out of date.

Reply:
Nice catch - I didn't think about this when I made the other change last night. This is trickier.

The reason we need this at all is that the RHS of assignments in the `try` block still might have "close" in their MC type. For example, in this code:
```
try (Socket s = new Socket(addr, port)) { } catch (Exception e) { }
```
`Socket s` (the LHS) is correctly assigned the type `@MustCall({})` by taking it's normal type and removing "close", because it's a try-with-resources variable. The RHS, however, has no indication that it's a try-with-resources variable; it's such a `new Socket` - which correctly has the type `@MustCall("close")`. Without this change, then we issue an incompatible assignment warning here.

I think what we actually need to check here is that if the LHS is a resource variable, then we should remove `"close"` from the RHS type and then do the check. I'll try to get that working.

Reply:
Thanks for the explanation.  Based on what you wrote, I am not sure if you need to make a further change here, other than updating the comment.  Can you give a test case where the initial assignment for a try-with-resources variable _should_ yield an error?  I would hope these variables must be (effectively) final, so they can't be re-assigned.

Reply:
> I would hope these variables must be (effectively) final, so they can't be re-assigned.

That's correct - the compiler inserts the `final` keyword IIRC.

> Can you give a test case where the initial assignment for a try-with-resources variable should yield an error?

What about this:
```
@MustCall({"close", "myMethod"})
class Foo extends Socket { void myMethod() {} /* also constructors etc */ }
try (Socket s = new Foo(addr, port)) { } catch (Exception e) { }
```

I'm actually not sure how much we care about a case like this, but I think at this point we should warn, because the LHS isn't `@MustCall("myMethod")`.

I think the fix is actually to do the subtyping check without close on both sides in the resource variable case. I think I can do that.

Reply:
I think it's fine to report an error at the initial assignment in this case for now.  Eventually I think the "right" fix would be to have the inferred type of `s` to be `@MustCall("myMethod")` and then to get rid of any special-casing of try-with-resources variables in `MustCallInvokedChecker`.  Then if `s.myMethod()` is invoked within the try block, the code will type check.

Reply:
The current code won't report an error.

I agree that's what the code *should* look like, and I think right now this will work correctly if the declared type of the resource variable is `Foo` rather than `Socket`.